vignettes/fims-path-maturity.Rmd

```diff
@@ -0,0 +1,549 @@
+
+Here, we are setting the maturity_models pointer in `Information` to equal the `lm0` pointer to the `LogisticMaturity` module
+```{Rcpp, eval = FALSE}
+/*
+ the maturity_models pointer from Information that matches the id of the 
+ fims::LogisticMaturity class is equal to the pointer to fims::LogisticMaturity 
+ */
+d0->maturity_models[lm0->id] = lm0;
+```
+
+```{r image-path-maturity-6, echo = FALSE, message=FALSE, out.width = '100%'}
+knitr::include_graphics("figures/fims-path-maturity-6.png")
+```
+
+Now we need to pass this pointer to the maturity pointer in population so that `population->maturity` points to the `LogisticMaturity` module instead of `MaturityBase`. First we set up a map in `Information` that points to `Population`. 
+```{Rcpp, eval = FALSE}
+std::map<uint32_t, std::shared_ptr<fims::Population<T> > >
+      populations; /*!<hash map to link each object to its shared location in
+                      memory*/
+  typedef typename std::map<uint32_t,std::shared_ptr<fims::Population<T> > >::iterator
+      population_iterator;
+  /**< iterator for population objects>*/
+```
+
+```{r image-path-maturity-7, echo = FALSE, message=FALSE, out.width = '100%'}
+knitr::include_graphics("figures/fims-path-maturity-7.png")
+```
+
+We then loop through the populations and create a new shared pointer, `p`, to reference the individual population of interest.  
+
+```{Rcpp, eval = FALSE}
+ for (population_iterator it = this->populations.begin();
+             it != this->populations.end(); ++it) {
+          std::shared_ptr<fims::Population<T> > p = (*it).second;
+          ...
+```
+In this code chunk, `(*it)` refers to a single population in the `populations` map. The second element of the map is the pointer, so `p = (*it).second` means we are setting the pointer to a single population to equal the pointer to `populations` in the map. 
+
+Within this population loop, we then link the `maturity` pointer in `population` to equal the `maturity` pointer in information, passing on the information that we want to use the `LogisticMaturity` class:
+
+```{Rcpp, eval = FALSE}  
+  // set maturity
+  if (p->maturity_id != -999) {
+    uint32_t maturity_uint = static_cast<uint32_t>(p->maturity_id);
+    maturity_models_iterator it = this->maturity_models.find(
+        maturity_uint);  // >maturity_models is specified in
+                         // information.hpp and used in rcpp
+
+    if (it != this->maturity_models.end()) {
+      p->maturity =
+          (*it).second;  // >maturity defined in population.hpp
+    }
+```
+
+Here, (*it) is referring to the `maturity_models` pointer.
+
+```{r image-path-maturity-8, echo = FALSE, message=FALSE, out.width = '100%'}
+knitr::include_graphics("figures/fims-path-maturity-8.png")
+```
+
+
+## Conceptualizing this path with R
```

Comment:
I like the idea, but this didn't work for me (I felt like I spent all that time trying to figure out the path in C++ and then had to start over to understand it in R), but I also don't always find analogies helpful. It might still be helpful for others!

Reply:
Thanks for the feedback! I added the following to teh intro of this section to help clarify it is optional and not required to understand the FIMS architecture!

This is an optional exercise for those who benefit from thinking through the structure in C++ code by seeing its corollary in R.